presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PickTableLayout.java

```diff
@@ -183,55 +205,157 @@ public Result apply(TableScanNode tableScanNode, Captures captures, Context cont
                 return Result.empty();
             }
 
-            return Result.ofPlanNode(planTableScan(tableScanNode, TRUE_LITERAL, context, metadata, domainTranslator));
+            return Result.ofPlanNode(planTableScan(tableScanNode, TRUE_LITERAL, context.getSession(), context.getSymbolAllocator().getTypes(), context.getIdAllocator(), metadata, parser, domainTranslator));
         }
     }
 
-    private static PlanNode planTableScan(TableScanNode node, Expression predicate, Rule.Context context, Metadata metadata, DomainTranslator domainTranslator)
+    private static PlanNode planTableScan(
+            TableScanNode node,
+            Expression predicate,
+            Session session,
+            TypeProvider types,
+            PlanNodeIdAllocator idAllocator,
+            Metadata metadata,
+            SqlParser parser,
+            DomainTranslator domainTranslator)
     {
+        return listTableLayout(
+                node,
+                predicate,
+                false,
+                session,
+                types,
+                idAllocator,
+                metadata,
+                parser,
+                domainTranslator)
+                .get(0);
+    }
+
+    public static List<PlanNode> listTableLayout(
+            TableScanNode node,
+            Expression predicate,
+            boolean pruneWithPredicateExpression,
+            Session session,
+            TypeProvider types,
+            PlanNodeIdAllocator idAllocator,
+            Metadata metadata,
+            SqlParser parser,
+            DomainTranslator domainTranslator)
+    {
+        // don't include non-deterministic predicates
         Expression deterministicPredicate = filterDeterministicConjuncts(predicate);
+
         DomainTranslator.ExtractionResult decomposedPredicate = DomainTranslator.fromPredicate(
                 metadata,
-                context.getSession(),
+                session,
                 deterministicPredicate,
-                context.getSymbolAllocator().getTypes());
+                types);
 
-        TupleDomain<ColumnHandle> simplifiedConstraint = decomposedPredicate.getTupleDomain()
+        TupleDomain<ColumnHandle> newDomain = decomposedPredicate.getTupleDomain()
                 .transform(node.getAssignments()::get)
                 .intersect(node.getCurrentConstraint());
 
+        Map<ColumnHandle, Symbol> assignments = ImmutableBiMap.copyOf(node.getAssignments()).inverse();
+
+        Constraint<ColumnHandle> constraint;
+        if (pruneWithPredicateExpression) {
```

Comment:
We do this because invoking the expression interpreter is currently very expensive (it used to be fast, fwiw). Add a comment explaining that and a TODO to get rid of this branch once we optimize the interpreter to make it cheap enough to call always.